### PR TITLE
Implementation of isCloseToUtcNow

### LINKED ISSUE #1519: Feature Request: assertThat(date).isCloseToUtcNow()
#### Summary

It is common that we have some code that creates a date-time (e.g. `LocalDateTime`) based on the current time, in UTC time zone.

We have a custom assertion for that, but it would be perhaps nice if assertJ had this out of the box.

This is our class:

```java
public final class DateAssertions {

    private static final int DELTA = 750;

    private DateAssertions() {}

    public static void assertNow(OffsetDateTime dateTime) {
        assertThat(dateTime).isCloseTo(OffsetDateTime.now(), within(DELTA, ChronoUnit.MILLIS));
    }

    public static void assertNow(LocalDateTime dateTime) {
        assertThat(dateTime).isCloseTo(LocalDateTime.now(Clock.systemUTC()), within(DELTA, ChronoUnit.MILLIS));
    }
}
```

We'd like to have the same functionality out of the box, e.g.

```java
assertThat(dateTime).isCloseToUtcNow();
```

perhaps with some overloads like:

```java
assertThat(dateTime).isCloseToUtcNow(DIFFERENT_DELTA, ChronoUnit.MILLIS);
```

If you also find this useful, I would be happy to create a PR for it.

### PR DESCRIPTION
**Pull request by ngeor**
#### Check List:
* Fixes #1519 
* Unit tests : YES
* Javadoc with a code example (API only) : YES

### CODE REVIEW

File: src/main/java/org/assertj/core/api/AbstractLocalDateTimeAssert.java [hunk 1]
```diff
@@ -317,6 +321,30 @@ public SELF isNotIn(String... dateTimesAsString) {
...
 
+  /**
+   * Verifies that the actual {@link LocalDateTime} is close to the current date and time on the UTC timezone,
+   * according to the given {@link TemporalUnitOffset}.
+   * You can build the offset parameter using {@link Assertions#within(long, TemporalUnit)} or {@link Assertions#byLessThan(long, TemporalUnit)}.
+   * <p>
+   * Example:
+   * <pre><code class='java'> LocalDateTime actual = LocalDateTime.now(Clock.systemUTC());
+   *
+   * // assertion will pass
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/AbstractLocalDateTimeAssert.java]: to be explicit let's add: `as it is executed less than one second after actual was built`

File: src/main/java/org/assertj/core/api/AbstractLocalDateTimeAssert.java [hunk 2]
```diff
@@ -317,6 +321,30 @@ public SELF isNotIn(String... dateTimesAsString) {
...
+   * assertThat(actual).isCloseToUtcNow(within(1, ChronoUnit.SECONDS));
+   *
+   * // assertion will fail
+   * assertThat(actual.plusSeconds(2)).isCloseToUtcNow(within(1, ChronoUnit.SECONDS));
+   * </code></pre>
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/AbstractLocalDateTimeAssert.java]: joins lines, this should look like:
```
   * assertThat(actual.plusSeconds(2)).isCloseToUtcNow(within(1, ChronoUnit.SECONDS));</code></pre>
```

File: src/main/java/org/assertj/core/api/AbstractLocalDateTimeAssert.java [hunk 3]
```diff
@@ -317,6 +321,30 @@ public SELF isNotIn(String... dateTimesAsString) {
...
+   * @param offset The offset used for comparison
+   * @return this assertion object
+   * @throws NullPointerException if {@code offset} parameter is {@code null}.
+   * @throws AssertionError if the actual {@code LocalDateTime} is {@code null}.
+   * @throws AssertionError if the actual {@code LocalDateTime} is not close to the current time for a provided offset.
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/AbstractLocalDateTimeAssert.java]: better wording for `is not close to the current time for a provided offset.` would be `is not close to the current time by less than the given offset.`

File: src/main/java/org/assertj/core/api/AbstractLocalDateTimeAssert.java [hunk 4]
```diff
@@ -317,6 +321,30 @@ public SELF isNotIn(String... dateTimesAsString) {
     return isNotIn(convertToLocalDateTimeArray(dateTimesAsString));
   }
 
+  /**
+   * Verifies that the actual {@link LocalDateTime} is close to the current date and time on the UTC timezone,
+   * according to the given {@link TemporalUnitOffset}.
+   * You can build the offset parameter using {@link Assertions#within(long, TemporalUnit)} or {@link Assertions#byLessThan(long, TemporalUnit)}.
+   * <p>
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/AbstractLocalDateTimeAssert.java]: Add that if the difference is = offset, the assertion succeeds

File: src/main/java/org/assertj/core/api/AbstractOffsetDateTimeAssert.java [hunk 5]
```diff
@@ -238,6 +243,30 @@ public SELF isAfter(String offsetDateTimeAsString) {
...
+   * // assertion will fail
+   * assertThat(actual.plusSeconds(2)).isCloseToUtcNow(within(1, ChronoUnit.SECONDS));
+   * </code></pre>
+   * @param offset The offset used for comparison
+   * @return this assertion object
+   * @throws NullPointerException if {@code offset} parameter is {@code null}.
+   * @throws AssertionError if the actual {@code OffsetDateTime} is {@code null}.
+   * @throws AssertionError if the actual {@code OffsetDateTime} is not close to the current time for a provided offset.
+   */
+  public SELF isCloseToUtcNow(TemporalUnitOffset offset) {
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/AbstractOffsetDateTimeAssert.java]: same comment as the other `isCloseToUtcNow`

File: src/test/java/org/assertj/core/api/localdatetime/LocalDateTimeAssert_isCloseToUtcNow_Test.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+ * @author Nikolaos Georgiou
+ */
+public class LocalDateTimeAssert_isCloseToUtcNow_Test extends LocalDateTimeAssertBaseTest {
+
+  @Test
+  public void test_isCloseToUtcNow_assertion_success() {
+    // WHEN
+    LocalDateTime now = LocalDateTime.now(Clock.systemUTC());
+    // THEN
+    assertThat(now).isCloseToUtcNow(within(1, ChronoUnit.SECONDS));
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/localdatetime/LocalDateTimeAssert_isCloseToUtcNow_Test.java]: minor: static import `SECONDS` for better fluency

File: src/test/java/org/assertj/core/api/localdatetime/LocalDateTimeAssert_isCloseToUtcNow_Test.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+import static org.assertj.core.util.FailureMessages.actualIsNull;
+
+/**
+ * @author Nikolaos Georgiou
+ */
+public class LocalDateTimeAssert_isCloseToUtcNow_Test extends LocalDateTimeAssertBaseTest {
+
+  @Test
+  public void test_isCloseToUtcNow_assertion_success() {
+    // WHEN
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/localdatetime/LocalDateTimeAssert_isCloseToUtcNow_Test.java]: this is a `GIVEN` section

Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/localdatetime/LocalDateTimeAssert_isCloseToUtcNow_Test.java]: This a `GIVEN`  more than a `WHEN`

File: src/test/java/org/assertj/core/api/localdatetime/LocalDateTimeAssert_isCloseToUtcNow_Test.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+  }
+
+  @Test
+  public void test_isCloseToUtcNow_assertion_fail() {
+    // WHEN
+    LocalDateTime now = LocalDateTime.now(Clock.systemUTC()).minusSeconds(2);
+    // THEN
+    assertThatExceptionOfType(AssertionError.class).isThrownBy(
+      () -> assertThat(now).isCloseToUtcNow(within(1, ChronoUnit.SECONDS))
+    ).withMessageContaining("within 1 Seconds but difference was 2 Seconds");
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/localdatetime/LocalDateTimeAssert_isCloseToUtcNow_Test.java]: prefer using `AssertioUtil.expectAssertionError`
```java
AssertionError error = expectAssertionError(() -> assertThat(now).isCloseToUtcNow(within(1, SECONDS));
// THEN
assertThat(error).hasMessageContaining("within 1 Seconds but difference was 2 Seconds");
```

Comment (joel-costigliola) [reply src/test/java/org/assertj/core/api/localdatetime/LocalDateTimeAssert_isCloseToUtcNow_Test.java]: same comment applies to the other tests that check errors/failures.

File: src/test/java/org/assertj/core/api/localdatetime/LocalDateTimeAssert_isCloseToUtcNow_Test.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+import static org.assertj.core.api.Assertions.within;
+import static org.assertj.core.util.FailureMessages.actualIsNull;
+
+/**
+ * @author Nikolaos Georgiou
+ */
+public class LocalDateTimeAssert_isCloseToUtcNow_Test extends LocalDateTimeAssertBaseTest {
+
+  @Test
+  public void test_isCloseToUtcNow_assertion_success() {
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/localdatetime/LocalDateTimeAssert_isCloseToUtcNow_Test.java]: rename to `should_pass_when_executed_within_time_offset` (no need to repeat test and isCloseToUtcNow)

File: src/test/java/org/assertj/core/api/localdatetime/LocalDateTimeAssert_isCloseToUtcNow_Test.java [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+    // WHEN
+    LocalDateTime now = LocalDateTime.now(Clock.systemUTC());
+    // THEN
+    assertThat(now).isCloseToUtcNow(within(1, ChronoUnit.SECONDS));
+  }
+
+  @Test
+  public void test_isCloseToUtcNow_assertion_fail() {
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/localdatetime/LocalDateTimeAssert_isCloseToUtcNow_Test.java]: rename to `should_fail_when_executed_after_time_offset`

**Comment by joel-costigliola** [comment]
@ngeor I have commented the PR, thanks for the work. The comments were made for `LocalDateTime` assertions but they also applies to `OffsetDateTime` assertions, I haven't duplicated them for brevity's sake.

**Comment by ngeor** [comment]
Hi @joel-costigliola thank you for the feedback, please have another look when you find the time, it should all be fixed now.

**Comment by joel-costigliola** [comment]
Integrated thanks @ngeor !

**Pull request closed without merging**